### PR TITLE
Implement invoices without an amount (i.e. any-amount invoice)

### PR DESCRIPTION
**Pull request by ZmnSCPxj**
Fixes: #534

### CODE REVIEW

**Comment by ZmnSCPxj** [comment]
Mostly complete but I still need to make a test specifically for paying amountless invoices.

File: wallet/wallet.h [hunk 1]
```diff
@@ -89,7 +89,7 @@ struct wallet_payment {
 	struct sha256 payment_hash;
 	enum wallet_payment_status status;
 	struct pubkey *destination;
-	u64 msatoshi;
+	u64 *msatoshi;
```
Comment (rustyrussell) [inline wallet/wallet.h]: I think this whole changeset is unnecessary.  Payments we're making must have a known value, AFAICT.  The other two patches stand alone.

File: lightningd/invoice.c [hunk 2]
```diff
@@ -298,7 +299,7 @@ static void json_add_invoices(struct json_result *response,
 		json_object_start(response, NULL);
 		json_add_string(response, "label", i->label);
 		json_add_hex(response, "rhash", &i->rhash, sizeof(i->rhash));
-		json_add_u64(response, "msatoshi", i->msatoshi);
+		json_add_u64(response, "msatoshi", *i->msatoshi);
 		json_add_bool(response, "complete", i->state == PAID);
```
Comment (rustyrussell) [inline lightningd/invoice.c]: This needs to be conditional, otherwise we segv when i->msatoshi is NULL.

File: lightningd/invoice.c [hunk 3]
```diff
@@ -364,7 +365,7 @@ static void json_delinvoice(struct command *cmd,
 	json_object_start(response, NULL);
 	json_add_string(response, "label", i->label);
 	json_add_hex(response, "rhash", &i->rhash, sizeof(i->rhash));
-	json_add_u64(response, "msatoshi", i->msatoshi);
+	json_add_u64(response, "msatoshi", *i->msatoshi);
```
Comment (rustyrussell) [inline lightningd/invoice.c]: This needs to be conditional, otherwise we segv when i->msatoshi is NULL.

File: lightningd/invoice.c [hunk 4]
```diff
@@ -81,7 +81,7 @@ static void tell_waiter(struct command *cmd, const struct invoice *paid)
 	json_object_start(response, NULL);
 	json_add_string(response, "label", paid->label);
 	json_add_hex(response, "rhash", &paid->rhash, sizeof(paid->rhash));
-	json_add_u64(response, "msatoshi", paid->msatoshi);
+	json_add_u64(response, "msatoshi", *paid->msatoshi);
```
Comment (rustyrussell) [inline lightningd/invoice.c]: This needs to be conditional, otherwise we segv when i->msatoshi is NULL.

File: lightningd/invoice.c [hunk 5]
```diff
@@ -81,7 +81,7 @@ static void tell_waiter(struct command *cmd, const struct invoice *paid)
...
 	json_add_bool(response, "complete", paid->state == PAID);
 	if (paid->state == PAID)
 		json_add_u64(response, "pay_index", paid->pay_index);
```
Comment (rustyrussell) [inline lightningd/invoice.c]: This is, in fact, always true here.  (in tell_waiter).  Probably should just assert paid->state == PAID at the top.

Maybe we should share this json-printing-invoice routine somewhere?

Also, looks like we screwed up sometime, and this routine is called in two loops; that should be fixed too, while you're poking around :)

File: wallet/wallet.h (hunk 1 above)
Comment (ZmnSCPxj) [reply wallet/wallet.h]: The problem is that a payment entry is created when an invoice is created.  (This seems strange to me. I think it makes more sense to create a payment entry when the invoice is actually paid).

In the case that the invoice is "any", what value should we put in the msatoshi field of the payment entry created while creating the invoice?

File: lightningd/invoice.c (hunk 5 above)
Comment (ZmnSCPxj) [reply lightningd/invoice.c]: Probably indeed better to factor out json-printing an invoice into common/json.c instead.

Re: two loops in `resolve_invoice`, hahahaha, looks like a rebase of waitanyinvoice and/or waitinvoice things I did a short while ago got borked, will make a separate pull request for that.

File: lightningd/invoice.c (hunk 2 above)
Comment (ZmnSCPxj) [reply lightningd/invoice.c]: toink! Yes, will fix.

**Comment by ZmnSCPxj** [comment]
@rustyrussell The `wallet_payment` structure is used for payment history, it seems.

However it seems strange that a payment entry is created when an invoice is created.

What is the intent of `wallet_payment` vs `invoice`?

From only the names, I imagine:

1.  An `invoice` is a conditional promise to deliver some good in exchange for getting paid the invoice amount.
2.  A `wallet_payment` is a historical fact that some transfer of value (i.e. a payment) occurred.

So it seems to me that `wallet_payment` creation should be done at:

1.  When we successfully pay (I believe this is already done in `json_pay`).
2.  When we successfully receive.

However currently we create a `wallet_payment` whenever an `invoice` is created.  Which (as mentioned at review) seems strange to me.

Why is a payment entry created when an invoice is created?

----

Also slightly tangential --- it is possible for the `invoice` RPC command to error but still create an invoice (which it will not return).  This occurs, if `wallet_payment_add` fails, since this error handling is done after the invoice is added to the database and in particular to the in-memory list.

**Comment by ZmnSCPxj** [comment]
The dual loops in `resolve_invoice` are correct.  One is for `waitanyinvoice` waiters, the other is for `waitinvoice`.

**Comment by ZmnSCPxj** [comment]
The pull request is now "complete", in the sense, that it fixes the issue as stated.

However there is still the question of how to record the actual value paid to us.  The possibilities are:

1.  `ALTER TABLE invoices ADD COLUMN msatoshis_paid INTEGER;`, `UPDATE TABLE invoices SET msatoshis_paid = msatoshis WHERE state=1;`, add corresponding `msatoshis_paid` nullable field to `struct invoice`, dump this field as appropriate, and modify `resolve_invoice` to set this field and db column correctly.
2.  Don't create a `wallet_payment` and corresponding entry in `payments` table during invoice creation.  Instead, create the `wallet_payment` when we actually receive the payment, and record the actual paid msatoshi in the `((struct wallet_payment*) p)->msatoshis` field.

--------

Also, for any-amount invoices, is it better to not have a `msatoshi` field in the JSON result output, or is it better to have a `"msatoshi": "any"` field instead?

File: wallet/wallet.h (hunk 1 above)
Comment (rustyrussell) [reply wallet/wallet.h]: Ah, pretty sure we should create it in response to an actual incoming payment. @cdecker?

Using the same type for outgoing and incoming payment is pretty ugly; as is the layering between the wallet and the rest of the code here.   I'd rather have the wallet code take over struct invoice and struct pay_command, but that's a bigger change.

**Comment by rustyrussell** [comment]
Ah, I see re: dual loops.  Yech, that's horribly subtle.  Renaming is a moral imperative here! :)

s/invs->invoice_waiters/waitany/ and s/invoice->invoice_waiters/waitone/ perhaps?

Better comment would help, too ofc.  I would do this, except it'll create gratuitous clash with your wwork...

File: wallet/wallet.h (hunk 1 above)
Comment (ZmnSCPxj) [reply wallet/wallet.h]: I could remove this changeset, and instead just load `payment.msatoshi = 0;` for any-amount invoices for now. Then we can open a separate issue for properly creating incoming payments when actual incoming payment occurs.

Comment (cdecker) [reply wallet/wallet.h]: The idea of this was to be a backing for the listing of payments, so we don't have to merge invoices, incoming payments and outgoing payments. It'd be rather unexpected for people to issue an invoice only to not see it in their list.

But I'm happy to defer the creation until the payment comes in, and not show invoiced payments in the transfer list.

**Comment by ZmnSCPxj** [comment]
Okay, I am now proposing to merge this as-is, including the change to `struct wallet_payment` to make `msatoshi` a nullable field.

1. We probably need more discussion regarding just how `wallet_payment` should work exactly.  Record everything in a single table? Separate invoices, incoming payments, and outgoing payments?
2. If we decide to record pending incoming payments, we still need to record the fact that a pending incoming payment is an any-amount incoming payment anyway (i.e. payment msatoshi should still be a nullable field).
3. We also need to decide how to record the actual value paid for an invoice, given that any-amount invoices are possible and that fixed-amount invoices allow overpayment anyway.

File: wallet/wallet.h (hunk 1 above)
Comment (rustyrussell) [reply wallet/wallet.h]: @cdecker I should have reviewed this harder: it's horribly confusing.  There's no payment yet!  You don't want this is a "listpayments" API...

From a user POV, outgoing and incoming payments are completely different creatures.  I'm not even sure why you'd want to see incoming payments at all, in fact.  AFAICT they're *always* immediately paid once we pay invoice.  So, invoices tell you what's paid.

**Pull request merged**